mllib/src/main/scala/org/apache/spark/ml/feature/LSH.scala

```diff
@@ -138,21 +139,31 @@ private[ml] abstract class LSHModel[T <: LSHModel[T]]
       // Limit the use of hashDist since it's controversial
       val hashDistUDF = udf((x: Seq[Vector]) => hashDistance(x, keyHash), DataTypes.DoubleType)
       val hashDistCol = hashDistUDF(col($(outputCol)))
-
-      // Compute threshold to get around k elements.
-      // To guarantee to have enough neighbors in one pass, we need (p - err) * N >= M
-      // so we pick quantile p = M / N + err
-      // M: the number of nearest neighbors; N: the number of elements in dataset
-      val relativeError = 0.05
-      val approxQuantile = numNearestNeighbors.toDouble / count + relativeError
       val modelDatasetWithDist = modelDataset.withColumn(distCol, hashDistCol)
-      if (approxQuantile >= 1) {
-        modelDatasetWithDist
+      // for a small dataset, use BoundedPriorityQueue
+      if (count < 1000) {
+        val queue = new BoundedPriorityQueue[Double](count.toInt)(Ordering[Double])
+        modelDatasetWithDist.collect().foreach { case Row(keys, values, distCol: Double) =>
+          queue += distCol
+        }
+        var sortedDistCol = queue.toArray.sorted(Ordering[Double])
+        queue.clear()
+        modelDatasetWithDist.filter(col(distCol) <= sortedDistCol(numNearestNeighbors - 1))
```

Comment:
I might pull `sortedDistCol(numNearestNeighbors - 1)` into a `val` or else this has to send the whole array. (You don't have to clear the queue)

Reply:
Will fix this. Thanks!